### DISCUSSION TITLE
Can Windows Forms support be added?

### DISCUSSION THREAD

**Discussion by lucinelouhi** [Mapsui/Mapsui#2613] [Ideas]
Support for Windows Forms can be increased. Because some existing platforms have high requirements for computer performance, and Windows Forms are also a common and fast way to send messages. Developers can focus on algorithms rather than beautifying the interface.

**Comment by pauldendulk** [comment]
It could, but it is not something that I will work on. And I don't know if you need it because you could also use our Eto.Forms option: https://www.nuget.org/packages/Eto.Forms

Considerations: One of our guidelines is, support less and support that better, but since we already have a lot of platforms adding another will not costs a lot of maintenance. We already have most of our code in shared projects. In this case we could probably also share most of the MapControl.cs code between Eto.Forms and Windows.Forms. 

There is a Windows.Forms SkiaSharp View so that is also not a problem https://www.nuget.org/packages/SkiaSharp.Views.WindowsForms/

btw, I was surprisded to see a Windows.Forms talk on .NET Conf, though it was dead https://learn.microsoft.com/en-us/shows/dotnetconf-2023/whats-new-with-winforms.

**Comment by lucinelouhi** [comment]
Thank you, I will use SkiaSharp Views SKControl in” SkiaSharp.Views.WindowsForms“ implements the IMapControl interface as the base class. It should be possible to use Mapsui in WinForm. Thank you very much for developing this basic library, which makes cross platform easier.

**Comment by charlenni** [comment]
You should do it like in the Maui implementation. If you use SKControl as base class, you always have a CPU supported control and can never change to a GPU supported control without recompiling Mapsui. And GPU support is much faster than CPU support.

That said, implementing IMapControl in a Panel and then add a SKControl or SKGLControl respecting the UseGPU flag would be the better way.

**Reply by lucinelouhi** [reply]
Thank you! I will refer to the implementation code of other platforms and try it out myself.

**Reply by charlenni** [reply]
It is possible. I checked it. There isn't any manupilation handling and PixelDensity is also missing, that should be possible to make. Would look for this into the Maui implementation. First shot looks like this: 
![grafik](https://github.com/Mapsui/Mapsui/assets/4381223/e8c36c74-d849-4861-8e1b-496359777dfb)
You could find it [here](https://github.com/charlenni/Mapsui/tree/WinForms).

**Reply by lucinelouhi** [reply]
Thank you, it's amazing

**Reply by lucinelouhi** [reply]
Did you use Maui directly to modify it?

**Reply by charlenni** [reply]
No, I did it from scratch. If I have time, I try to look into manipulation of the control and perhaps to implement the normal sample behavior.

**Reply by lucinelouhi** [reply]
Amazing! I have to experiment slowly. I can't do anything if my level is not high. Can you give me the code you implemented for my reference? Thank you. lucinelouhi@gmail.com

**Reply by lucinelouhi** [reply]
Thank you! I saw that code.

**Reply by charlenni** [reply]
Ok, MapControl for WinForms is now ready. The same with the samples app. You now could test the different samples, which Mapsui has. 

I found up to now two problems:
1. There is an error when the app is started. Seems to be a WinForms problem and I don't know WinForms.
2. If you rotate the map via the Navigator, then the rotation track bar isn't updated. The ViewportChanged event isn't raised. Not logical to me, why.

**Comment by charlenni** [comment]
Ok, also these two problems are solved. The TrackBar is still a problem, because it could only use integers, but Viewport.Rotation is double. This is not a problem of Mapsui.WindowsForms, but from TrackBar. But I don't want to add any other control for this.

**Reply by inforithmics** [reply]
@charlenni Do you plan to make a pull request to Add Mapsui.WindowsForms back to Mapsui?

**Reply by charlenni** [reply]
No, I will not. Paul said, that I make bad code, waste his time and should keep away from Mapsui.

I'm still here, because after 7 years it need some time to leave a project.

**Reply by charlenni** [reply]
But you can grab it and make a PR of it, if you think it's good enough and useful. If not, the interested devs could grab it by themselves.

**Reply by inforithmics** [reply]
In my opinion you write fantastic code and are very knowledgeable and don't waste my time. So I will look at the code and see if I make a pull request myself (with the history preserved).

**Reply by charlenni** [reply]
Thank you , Thomas, for your warm words. I don't have your knowledge about tolls and workflow, but always hoped to contribute something helpful.

There are some more things I worked on the last time. One is an improvement of the MAUI and Avalonia sample. I added to both of the a slider for the rotation and a layer list, so that it is possible to enable and disable each individual layer. I couldn't do it for Uno and Eto, because I don't know anything of them. Perhaps there is someone using them and could add this functionality. See https://github.com/charlenni/Mapsui/tree/add-layerlist.

The second I thought about was a better MemoryProvider to get rid of MemoryLayer and WritableLayers. There isn't a clear line about what is responsible for what. Normally a Layer should provide the Styles and the Provider the Features. A MemoryLayer is then a combination of two things. And what is the idea behind a WritableLayer? I implemented the IList interface for the MemoryProvider. I’m still working on this.

The third is the so bad made MapView replacement. MapView was always a convenience layer around the MapControl. So I tried to replace this things with extensions to the normal objects. You can find the result at https://github.com/charlenni/Mapsui/tree/add-markerextension

Best regards,
Dirk

**Reply by charlenni** [reply]
@inforithmics I now uploaded a first version of the extended MemoryProvider, that should replace MemoryLayer and WritableLayer. It isn't perfect and the removing are not done. I have to do this later.

You could find the code [here](https://github.com/charlenni/Mapsui/tree/change-memorylayer).

**Comment by inforithmics** [comment]
Here is the pull request https://github.com/Mapsui/Mapsui/pull/2690 (Learned some git tricks along the way).

**Comment by inforithmics** [comment]
Thanks for your work I will have a look at the Branches and eventually make more pull requests.

**Reply by charlenni** [reply]
Thomas, there is no reason to say thank you for my work. You and I both interested to support Mapsui in different ways. I think, that this is a good thing to do.

It would be nice, if you could have a look into the code. Perhaps there is something helpful that you can use.
